### PR TITLE
[TDL-25607] : API version upgrade

### PR DESCRIPTION
**Pull request by shantanu73**
# Description of change
Klaviyo is moving it's legacy APIs v1/v2 to new APIs as per [this document](https://developers.klaviyo.com/en/docs/apis_comparison_chart).
The following changes are made to include these API changes:
1. Changed the base url for all streams
2. Moved API key authorization from query params to request headers
3. Added new property api_version in tap config.
4. Modified schema for all streams to include all relevant fields.
5. Modified transformation logic to flatten the API response and get all relevant fields.

# Manual QA steps
 - Ran syncs for customer connection
 
# Risks
 - Old API implementation is getting deprecated and it'll sunset soon, will cause connections to fail.
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_klaviyo/__init__.py [hunk 1]
```diff
@@ -172,29 +174,32 @@ def get_available_metrics(api_key):
...
+def do_discover(headers):
+    print(json.dumps(discover(headers), indent=2))
 
 @singer.utils.handle_top_exception(LOGGER)
 def main():
 
     args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)
+    headers = {
+        "Authorization": f"Klaviyo-API-Key {args.config.get('api_key')}",
+        "revision": f"{args.config.get('api_version', API_VERSION)}"
```
Comment (rdeshmukh15) [inline tap_klaviyo/__init__.py:192]: AFAIK, API_VERSION is hardcoded value in the code and should not be taken from the config.

File: tap_klaviyo/__init__.py [hunk 2]
```diff
@@ -5,37 +5,40 @@
...
 
+API_VERSION = "2024-02-15"
+
 ENDPOINTS = {
-    'global_exclusions': 'https://a.klaviyo.com/api/v1/people/exclusions',
-    'lists': 'https://a.klaviyo.com/api/v1/lists',
-    'metrics': 'https://a.klaviyo.com/api/v1/metrics',
-    'metric': 'https://a.klaviyo.com/api/v1/metric/',
-    'campaigns': 'https://a.klaviyo.com/api/v1/campaigns'
+    'global_exclusions': 'https://a.klaviyo.com/api/profiles',
```
Comment (rdeshmukh15) [inline tap_klaviyo/__init__.py:15]: Can you please provide the document reference to determin this API endpoint replacement?

File: config.sample.json [hunk 3]
```diff
@@ -1,5 +1,6 @@
 {
     "api_key": "asasdkldasdlkaasdkfdsjadf",
+    "api_version": "2024-02-15",
```
Comment (rdeshmukh15) [inline config.sample.json:3]: api_version should not be the configurable property unless your code is flexible enough to switch between different api versions.

File: README.md [hunk 4]
```diff
@@ -29,6 +29,7 @@ Singer taps function in two modes: [discovery mode](https://github.com/singer-io
     ```json
     {
         "api_key": "pk_XYZ",
+        "api_version": "2024-02-15",
```
Comment (rdeshmukh15) [inline README.md:32]: Remove api_version from config.json as api_version should not be the configurable property unless your code is flexible enough to switch between different api versions.

**Comment by rdeshmukh15** [comment]
Klaviyo metrics API link should be updated to the latest in README.md

**Comment by rdeshmukh15** [comment]
Have you compared the synced data for old and new both the versions with same start_date?

File: config.sample.json (hunk 3 above)
Comment (shantanu73) [reply config.sample.json:3]: Sure, I'll make that change.

File: README.md (hunk 4 above)
Comment (shantanu73) [reply README.md:32]: Sure, I'll make that change.

File: tap_klaviyo/__init__.py (hunk 2 above)
Comment (shantanu73) [reply tap_klaviyo/__init__.py:15]: Yes, I'll do that.

File: tap_klaviyo/__init__.py (hunk 1 above)
Comment (shantanu73) [reply tap_klaviyo/__init__.py]: Fixed it.

**Review by rdeshmukh15 — APPROVED** [review]

File: tap_klaviyo/utils.py [hunk 5]
```diff
@@ -179,68 +179,113 @@ def authed_get(source, url, params):
...
+                    relationship_data = [relationship_data]
+                for relationship in relationship_data:
+                    # Flatten the relationship dict with data
+                    relationship_id = relationship['id']
+                    for included_relationship in included:
+                        # Break out of the loop if included relationship is found
+                        if relationship_id == included_relationship['id']:
+                            # Flatten the included_relationship dict with attributes
+                            included_relationship.update(included_relationship['attributes'])
+                            relationship.update(included_relationship)
```
Comment (Vi6hal) [inline tap_klaviyo/utils.py:285]: use a mapping / dict to fetch the related relationship in the included list, as going through the whole list for each event will be time consuming.

File: tap_klaviyo/utils.py [hunk 6]
```diff
@@ -179,68 +179,113 @@ def authed_get(source, url, params):
...
+                    "additional-fields[profile]": "subscriptions,predictive_analytics"
+                }
+            ]
+        else:
+            # params list for lists stream 
+            params_list = [
+                {
+                    "include": "tags"
+                }
+            ]
```
Comment (Vi6hal) [inline tap_klaviyo/utils.py:249]: ## Optional
Move these params list to the class declaration or to a dict mapping of the stream so this block will not be required

Comment (shantanu73) [reply tap_klaviyo/utils.py]: Done

File: tap_klaviyo/utils.py (hunk 5 above)
Comment (shantanu73) [reply tap_klaviyo/utils.py]: Done

**Review by Vi6hal — APPROVED** [review]

**Review by luandy64 — APPROVED** [review]

**Pull request merged**